clang/test/SemaSYCL/esimd-sycl-context-switch.cpp

```diff
@@ -0,0 +1,27 @@
+// RUN: %clang_cc1 -fsycl-is-device -fsyntax-only -verify %s
+
+// This test checks that SYCL device functions cannot be called from ESIMD context.
+
+__attribute__((sycl_device)) void sycl_func() {}
+__attribute__((sycl_device)) void __reserved_func() {}
+
+// Immediate diagnostic
+__attribute__((sycl_device)) __attribute__((sycl_explicit_simd)) void esimd_func1() {
+  // expected-error@+1{{SYCL device function cannot be called from ESIMD context}}
+  sycl_func();
+  // Reserved functions are allowed
+  __reserved_func();
+}
+
+// Deffered diagnostic
+void foo() {
+  // expected-error@+1{{SYCL device function cannot be called from ESIMD context}}
+  sycl_func();
+  // Reserved functions are allowed
+  __reserved_func();
+}
+
+__attribute__((sycl_device)) __attribute__((sycl_explicit_simd)) void esimd_func2() {
+  // expected-note@+1{{called by}}
+  foo();
+}
```

Comment:
Some other tests that could be interesting:
```
struct S {
  __attribute__((sycl_device)) void sycl_func() {}
};

__attribute__((sycl_device)) __attribute__((sycl_explicit_simd)) void esimd_func1() {
  S s;
  s.sycl_func();
}
```
```
template <typename Ty>
__attribute__((sycl_device)) void sycl_func() {}

__attribute__((sycl_device)) __attribute__((sycl_explicit_simd)) void esimd_func1() {
  sycl_func<int>();
}
```
```
namespace std {
template <typename _Tp>
_Tp declval();

template <typename _Functor, typename... _ArgTypes>
struct __res {
  template <typename... _Args>
  static decltype(declval<_Functor>()(_Args()...)) _S_test(int);

  template <typename...>
  static void _S_test(...);

  typedef decltype(_S_test<_ArgTypes...>(0)) type;
};

template <typename>
struct function;

template <typename _R, typename... _ArgTypes>
struct function<_R(_ArgTypes...)> {
  template <typename _Functor,
            typename = typename __res<_Functor, _ArgTypes...>::type>
  __attribute__((sycl_device, sycl_explicit_simd)) function(_Functor) {}
  __attribute__((sycl_device, sycl_explicit_simd)) _R operator()(_ArgTypes...) const;
};
} // namespace std

__attribute__((sycl_device)) void sycl_func() {}

__attribute__((sycl_device, sycl_explicit_simd)) void passthrough(std::function<void(void)> &&C) { C(); }

__attribute__((sycl_device)) __attribute__((sycl_explicit_simd)) void esimd_func1() {
  passthrough(sycl_func);
}
```

Reply:
I checked all 3 cases. For all of them, the error is emitted as expected. Let me know if you want those cases to be included.

Reply:
I would appreciate you adding them -- it helps prevent regressions in the future.

Reply:
Done